eskip/template.go

```diff
@@ -39,15 +41,54 @@ func NewTemplate(template string) *Template {
 // Apply evaluates the template using a TemplateGetter function to resolve the
 // placeholders.
 func (t *Template) Apply(get TemplateGetter) string {
-	result := t.template
-
 	if get == nil {
-		return result
+		return t.template
 	}
+	result, _ := t.apply(get)
+	return result
+}
 
-	for _, placeholder := range t.placeholders {
-		result = strings.Replace(result, "${"+placeholder+"}", get(placeholder), -1)
+// ApplyRequestContext evaluates the template using a filter context and request attributes to resolve the
+// placeholders. Returns true if all placeholders resolved to non-empty values.
+func (t *Template) ApplyRequestContext(ctx filters.FilterContext) (string, bool) {
+	return t.apply(requestGetter(ctx))
+}
+
+// ApplyResponseContext evaluates the template using a filter context, request and response attributes to resolve the
+// placeholders. Returns true if all placeholders resolved to non-empty values.
+func (t *Template) ApplyResponseContext(ctx filters.FilterContext) (string, bool) {
+	return t.apply(responseGetter(ctx))
+}
+
+func requestGetter(ctx filters.FilterContext) func(key string) string {
+	return func(key string) string {
+		if v := ctx.PathParam(key); v != "" {
+			return v
+		}
+		if v, ok := ctx.StateBag()[key]; ok {
```

Comment:
What do you think of supporting here additional primitve types? Basically only: integers and floats.

Reply:
I have updated PR to support numbers from statebag.

TBH I have a bit of a doubt regarding statebag here, maybe we can discuss it.
**In short**: do we actually need to give access to state bag at all?

Within #1444 I initially thought about using statebag as a kind of buffer between filter that copies e.g. header to statebag (e.g. `getRequestHeader("X-Foo-Header", "foo")`) and filter that uses stored value to e.g. set header (e.g `setRequestHeader("X-Bar-Header", "${foo}")`). With this in mind the ticket was outlined and initial templating #1462 implemented. 

Some time later the _projection_ idea [came up](https://github.com/zalando/skipper/issues/1444#issuecomment-700566455) - use placeholder prefix to denote datasource, e.g. `${request.header.X-Foo-Header}`, which I like and plan as a [subsequent extension](https://github.com/zalando/skipper/issues/1444#issuecomment-700967310) to this.

Prefixed placeholders obviate the need to have filters to store values into statebag.

`PathParam` placeholder replacement was done in #1462  with this PR (unified templating across filters) in mind - path params are resolved without a prefix to stay compatible with already existing `setQuery` and `setPath`.

#1462 added path params and statebag resolution but did not touch template regexp
https://github.com/zalando/skipper/blob/79de18de90fd338afae2a6ca6ac3e3bac8c46741/eskip/template.go#L12
which only allows words as placeholders, so actually many existing statebag keys like `auth-user` or `backend:isproxy` can not be used unless placeholder regexp is updated (which is necessary to support prefixed placeholders) - `${auth-user}` or `${backend:isproxy}` are not recognized as placeholders.

Since we did not hear complains from users about non-word placeholders, as well as, there are no means to store something into statebag in eskip we have a chance to retract and remove statebag support from templating. And maybe introduce it later as a prefixed placeholder `${statebag.whatever}` if needed. 

I just don't like two things: stagebag has internal use to pass data between skipper components (filters, proxy) and this PR exposes it without prefix (because it was done without prefix initially in #1462)
Or maybe I've just overthought all of this and having primitives from stagebag without a prefix is fine.

Reply:
i like this approach, and I think the templating could work well without the statebag. :+1: